[GTK][WPE] Don't allow depth test and stencil clipping if packed depth stencil is not supported

magomez opened the pull request:
#### ed286350cfd576f534f27200fa06d714fb14f8a9
<pre>
[GTK][WPE] Don&apos;t allow depth test and stencil clipping if packed depth stencil is not supported
<a href="https://bugs.webkit.org/show_bug.cgi?id=273177">https://bugs.webkit.org/show_bug.cgi?id=273177</a>

Reviewed by Carlos Garcia Campos.

Use a packed depth stencil format when available, which allows depth testing
and stencil clipping at the same time. If packed depth stencil is not
supported, use separate buffers for depth and stencil. In the latter case,
if both buffers are requested at the same time, the stencil buffer won&apos;t
be created.

* Source/WebCore/platform/graphics/egl/GLContext.cpp:
(WebCore::GLContext::glExtensions const):
* Source/WebCore/platform/graphics/egl/GLContext.h:
* Source/WebCore/platform/graphics/texmap/BitmapTexture.cpp:
(WebCore::depthBufferFormat):
(WebCore::BitmapTexture::initializeStencil):
(WebCore::BitmapTexture::initializeDepthBuffer):
(WebCore::BitmapTexture::~BitmapTexture):
* Source/WebCore/platform/graphics/texmap/BitmapTexture.h:

Canonical link: <a href="https://commits.webkit.org/278159@main">https://commits.webkit.org/278159@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a1b15109e2c31c310ffd388b10bd743aa6f5203c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49529 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28811 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/52573 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/52768 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/202 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/51833 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/34829 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/26431 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40442 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/51629 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/26353 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/42675 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21557 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23814 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/43858 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/7898 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/45728 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/44363 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54344 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24613 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/20803 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/47812 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/25883 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/42834 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/46829 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/26725 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7148 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/25606 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fujii commented on Source/WebCore/platform/graphics/texmap/BitmapTexture.cpp:
```diff
@@ -51,8 +51,34 @@ static const GLenum s_pixelDataType = GL_UNSIGNED_INT_8_8_8_8_REV;
 
+GLenum depthBufferFormat()
+{
+    static std::once_flag onceFlag;
```
TextureMapper is thread agnostic. Clients have to call all TextureMapper API from a single thread.
I think TextureMapper doesn't need thread synchronization.

magomez replied:
Yes, I know that TextureMapper is only used from the compositor thread. But I'm using std::call_once to ensure that the initialization of the depth buffer format happens only once. I could use a boolean flag to check whether the value is initialized or not, but std::call_once seems to be commonly used for that.

carlosgcampos replied:
I don't think it's worth calling this only once, since both the extension and the version are cached.

magomez committed:
[GTK][WPE] Don't allow depth test and stencil clipping if packed depth stencil is not supported
https://bugs.webkit.org/show_bug.cgi?id=273177

Reviewed by Carlos Garcia Campos.

Use a packed depth stencil format when available, which allows depth testing
and stencil clipping at the same time. If packed depth stencil is not
supported, use separate buffers for depth and stencil. In the latter case,
if both buffers are requested at the same time, the stencil buffer won't
be created.

* Source/WebCore/platform/graphics/egl/GLContext.cpp:
(WebCore::GLContext::glExtensions const):
* Source/WebCore/platform/graphics/egl/GLContext.h:
* Source/WebCore/platform/graphics/texmap/BitmapTexture.cpp:
(WebCore::depthBufferFormat):
(WebCore::BitmapTexture::initializeStencil):
(WebCore::BitmapTexture::initializeDepthBuffer):
(WebCore::BitmapTexture::~BitmapTexture):
* Source/WebCore/platform/graphics/texmap/BitmapTexture.h:

Canonical link: https://commits.webkit.org/278159@main
